# Machine Learning - Continuous Assessment - Class Group: TU060 <a class="tocSkip">

# MSc in Computer Science - Data Science (Part Time) <a class="tocSkip">

# Student Name: Ciaran Finnegan  <a class="tocSkip">

# Student No: D21124026 <a class="tocSkip">

#### Student EMail: D21124026@mytudublin.ie <a class="tocSkip">

## May 2022<a class="tocSkip">

# 

# H&M Personalised Fashion Recommendations <a class="tocSkip">


#### This project takes the datasets provided by Kaggle for the H&M Personalised Fashion Recommendations competition.  <a class="tocSkip">

Ipso forum...

## <a class="tocSkip">

#### The CA follows a n-part structure. <a class="tocSkip">

Section 1: Ipso forum...

Section 2: Ipso forum...

Section 3: Ipso forum...


### <a class="tocSkip">

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Jupyter-Notebook-Set-Up" data-toc-modified-id="Jupyter-Notebook-Set-Up-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Jupyter Notebook Set Up</a></span><ul class="toc-item"><li><span><a href="#Import-Libraries" data-toc-modified-id="Import-Libraries-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Import Libraries</a></span><ul class="toc-item"><li><span><a href="#General-Utility-Libraries" data-toc-modified-id="General-Utility-Libraries-1.1.1"><span class="toc-item-num">1.1.1&nbsp;&nbsp;</span>General Utility Libraries</a></span></li><li><span><a href="#Panda-+-Numpy" data-toc-modified-id="Panda-+-Numpy-1.1.2"><span class="toc-item-num">1.1.2&nbsp;&nbsp;</span>Panda + Numpy</a></span></li><li><span><a href="#Matplotlib-Library" data-toc-modified-id="Matplotlib-Library-1.1.3"><span class="toc-item-num">1.1.3&nbsp;&nbsp;</span>Matplotlib Library</a></span></li><li><span><a href="#Plotly-Libraries" data-toc-modified-id="Plotly-Libraries-1.1.4"><span class="toc-item-num">1.1.4&nbsp;&nbsp;</span>Plotly Libraries</a></span></li></ul></li></ul></li><li><span><a href="#Project-Data-Import,-Analysis-and-Preparation" data-toc-modified-id="Project-Data-Import,-Analysis-and-Preparation-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Project Data Import, Analysis and Preparation</a></span><ul class="toc-item"><li><span><a href="#Load-and-Sample-Kaggle-H&amp;M-Datasets" data-toc-modified-id="Load-and-Sample-Kaggle-H&amp;M-Datasets-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Load and Sample Kaggle H&amp;M Datasets</a></span><ul class="toc-item"><li><span><a href="#Set-Up-File-Paths" data-toc-modified-id="Set-Up-File-Paths-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>Set Up File Paths</a></span></li><li><span><a href="#H&amp;M-Transaction-Dataset" data-toc-modified-id="H&amp;M-Transaction-Dataset-2.1.2"><span class="toc-item-num">2.1.2&nbsp;&nbsp;</span>H&amp;M Transaction Dataset</a></span></li><li><span><a href="#H&amp;M-Articles-Dataset" data-toc-modified-id="H&amp;M-Articles-Dataset-2.1.3"><span class="toc-item-num">2.1.3&nbsp;&nbsp;</span>H&amp;M Articles Dataset</a></span></li><li><span><a href="#H&amp;M-Customers-Dataset" data-toc-modified-id="H&amp;M-Customers-Dataset-2.1.4"><span class="toc-item-num">2.1.4&nbsp;&nbsp;</span>H&amp;M Customers Dataset</a></span></li></ul></li><li><span><a href="#Exploratory-Data-Analysis" data-toc-modified-id="Exploratory-Data-Analysis-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Exploratory Data Analysis</a></span><ul class="toc-item"><li><span><a href="#H&amp;M-Sampled-Transaction-Dataset" data-toc-modified-id="H&amp;M-Sampled-Transaction-Dataset-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>H&amp;M Sampled Transaction Dataset</a></span></li><li><span><a href="#H&amp;M-Sampled-Customer-Dataset" data-toc-modified-id="H&amp;M-Sampled-Customer-Dataset-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>H&amp;M Sampled Customer Dataset</a></span></li><li><span><a href="#H&amp;M-Sampled-Articles-Dataset" data-toc-modified-id="H&amp;M-Sampled-Articles-Dataset-2.2.3"><span class="toc-item-num">2.2.3&nbsp;&nbsp;</span>H&amp;M Sampled Articles Dataset</a></span></li></ul></li><li><span><a href="#Data-Preparation-and-Feature-Enrichment" data-toc-modified-id="Data-Preparation-and-Feature-Enrichment-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Data Preparation and Feature Enrichment</a></span><ul class="toc-item"><li><span><a href="#Transaction-Data-Feature-Enrichment-(Sample)" data-toc-modified-id="Transaction-Data-Feature-Enrichment-(Sample)-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>Transaction Data Feature Enrichment (Sample)</a></span></li><li><span><a href="#Extend-Dataframe-with-Customer-Data" data-toc-modified-id="Extend-Dataframe-with-Customer-Data-2.3.2"><span class="toc-item-num">2.3.2&nbsp;&nbsp;</span>Extend Dataframe with Customer Data</a></span></li><li><span><a href="#Extend-Dataframe-with-Articles-Data" data-toc-modified-id="Extend-Dataframe-with-Articles-Data-2.3.3"><span class="toc-item-num">2.3.3&nbsp;&nbsp;</span>Extend Dataframe with Articles Data</a></span></li><li><span><a href="#Final-Data-Preparation-Steps" data-toc-modified-id="Final-Data-Preparation-Steps-2.3.4"><span class="toc-item-num">2.3.4&nbsp;&nbsp;</span>Final Data Preparation Steps</a></span></li></ul></li></ul></li><li><span><a href="#Split-Data:-Train-and-Test-Sets" data-toc-modified-id="Split-Data:-Train-and-Test-Sets-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Split Data: Train and Test Sets</a></span></li><li><span><a href="#Model-Preparation" data-toc-modified-id="Model-Preparation-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Model Preparation</a></span><ul class="toc-item"><li><span><a href="#Set-Up-the-Scaling-Algorithm" data-toc-modified-id="Set-Up-the-Scaling-Algorithm-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Set Up the Scaling Algorithm</a></span></li><li><span><a href="#Set-Up-the-Columns-to-be-Scaled" data-toc-modified-id="Set-Up-the-Columns-to-be-Scaled-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Set Up the Columns to be Scaled</a></span></li><li><span><a href="#Scale-the-Training-Dataset" data-toc-modified-id="Scale-the-Training-Dataset-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Scale the Training Dataset</a></span></li></ul></li><li><span><a href="#K-Nearest-Neighbour-Operations" data-toc-modified-id="K-Nearest-Neighbour-Operations-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>K Nearest Neighbour Operations</a></span><ul class="toc-item"><li><span><a href="#Create-Functions-for-NN-Analysis" data-toc-modified-id="Create-Functions-for-NN-Analysis-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Create Functions for NN Analysis</a></span><ul class="toc-item"><li><span><a href="#Set-up-function-to-Build-Nearest-Neighbours-Model" data-toc-modified-id="Set-up-function-to-Build-Nearest-Neighbours-Model-5.1.1"><span class="toc-item-num">5.1.1&nbsp;&nbsp;</span>Set up function to Build Nearest Neighbours Model</a></span></li><li><span><a href="#Set-up-function-to--Nearest-Neighbours-(+-Articles-bought)-for-a-Customer" data-toc-modified-id="Set-up-function-to--Nearest-Neighbours-(+-Articles-bought)-for-a-Customer-5.1.2"><span class="toc-item-num">5.1.2&nbsp;&nbsp;</span>Set up function to  Nearest Neighbours (+ Articles bought) for a Customer</a></span></li><li><span><a href="#Find-Top-Selling-Products-in-'Neighbourhood'-for-a-Customer" data-toc-modified-id="Find-Top-Selling-Products-in-'Neighbourhood'-for-a-Customer-5.1.3"><span class="toc-item-num">5.1.3&nbsp;&nbsp;</span>Find Top Selling Products in 'Neighbourhood' for a Customer</a></span></li><li><span><a href="#Invoke-Function-to-Build-KNN-Model" data-toc-modified-id="Invoke-Function-to-Build-KNN-Model-5.1.4"><span class="toc-item-num">5.1.4&nbsp;&nbsp;</span>Invoke Function to Build KNN Model</a></span></li></ul></li></ul></li><li><span><a href="#Evaluate-Model" data-toc-modified-id="Evaluate-Model-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Evaluate Model</a></span><ul class="toc-item"><li><span><a href="#Define-Functions-to-Extract-Customer-IDs-and-Article-Predictions-from-Test-Data" data-toc-modified-id="Define-Functions-to-Extract-Customer-IDs-and-Article-Predictions-from-Test-Data-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Define Functions to Extract Customer IDs and Article Predictions from Test Data</a></span><ul class="toc-item"><li><span><a href="#Define-Function-to-read-Customer-IDs-from-Submission-File" data-toc-modified-id="Define-Function-to-read-Customer-IDs-from-Submission-File-6.1.1"><span class="toc-item-num">6.1.1&nbsp;&nbsp;</span>Define Function to read Customer IDs from Submission File</a></span></li><li><span><a href="#Define-Function-to-Obtain-Customer-Transactions" data-toc-modified-id="Define-Function-to-Obtain-Customer-Transactions-6.1.2"><span class="toc-item-num">6.1.2&nbsp;&nbsp;</span>Define Function to Obtain Customer Transactions</a></span></li><li><span><a href="#Define-Function-to-Obtain-Customer-Features" data-toc-modified-id="Define-Function-to-Obtain-Customer-Features-6.1.3"><span class="toc-item-num">6.1.3&nbsp;&nbsp;</span>Define Function to Obtain Customer Features</a></span></li><li><span><a href="#Define-Function-to-Obtain-Article-Features" data-toc-modified-id="Define-Function-to-Obtain-Article-Features-6.1.4"><span class="toc-item-num">6.1.4&nbsp;&nbsp;</span>Define Function to Obtain Article Features</a></span></li><li><span><a href="#Define-Function-to-Obtain-Article-Predictions" data-toc-modified-id="Define-Function-to-Obtain-Article-Predictions-6.1.5"><span class="toc-item-num">6.1.5&nbsp;&nbsp;</span>Define Function to Obtain Article Predictions</a></span></li><li><span><a href="#Define-Functions-to-Parse-Article-Predictions-for-Kaggle-Submission" data-toc-modified-id="Define-Functions-to-Parse-Article-Predictions-for-Kaggle-Submission-6.1.6"><span class="toc-item-num">6.1.6&nbsp;&nbsp;</span>Define Functions to Parse Article Predictions for Kaggle Submission</a></span></li></ul></li><li><span><a href="#Define-Functions-to-Evaluate-Model-Predictions-aginst-Test-Data" data-toc-modified-id="Define-Functions-to-Evaluate-Model-Predictions-aginst-Test-Data-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Define Functions to Evaluate Model Predictions aginst Test Data</a></span><ul class="toc-item"><li><span><a href="#Set-up-function-to-return-latest-Customer-transactions-in-the-Test-Data" data-toc-modified-id="Set-up-function-to-return-latest-Customer-transactions-in-the-Test-Data-6.2.1"><span class="toc-item-num">6.2.1&nbsp;&nbsp;</span>Set up function to return latest Customer transactions in the Test Data</a></span></li><li><span><a href="#Define-Function-to-Return-Evaluation-Score" data-toc-modified-id="Define-Function-to-Return-Evaluation-Score-6.2.2"><span class="toc-item-num">6.2.2&nbsp;&nbsp;</span>Define Function to Return Evaluation Score</a></span></li><li><span><a href="#Iterate-Through-Test-Data-and-Generate-Predictions" data-toc-modified-id="Iterate-Through-Test-Data-and-Generate-Predictions-6.2.3"><span class="toc-item-num">6.2.3&nbsp;&nbsp;</span>Iterate Through Test Data and Generate Predictions</a></span></li></ul></li></ul></li><li><span><a href="#Build-Submission" data-toc-modified-id="Build-Submission-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Build Submission</a></span><ul class="toc-item"><li><span><a href="#Read-Sample-Submission-File" data-toc-modified-id="Read-Sample-Submission-File-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>Read Sample Submission File</a></span></li><li><span><a href="#Iterate-Through-Submission-File-to-Generate-Article-Predictions" data-toc-modified-id="Iterate-Through-Submission-File-to-Generate-Article-Predictions-7.2"><span class="toc-item-num">7.2&nbsp;&nbsp;</span>Iterate Through Submission File to Generate Article Predictions</a></span></li><li><span><a href="#Write-Predictions-Submission-File-to-CSV" data-toc-modified-id="Write-Predictions-Submission-File-to-CSV-7.3"><span class="toc-item-num">7.3&nbsp;&nbsp;</span>Write Predictions Submission File to CSV</a></span><ul class="toc-item"><li><span><a href="#SetUp-Disk-Directory-for-Submission-File" data-toc-modified-id="SetUp-Disk-Directory-for-Submission-File-7.3.1"><span class="toc-item-num">7.3.1&nbsp;&nbsp;</span>SetUp Disk Directory for Submission File</a></span></li><li><span><a href="#Create-Submission-.CSV-File" data-toc-modified-id="Create-Submission-.CSV-File-7.3.2"><span class="toc-item-num">7.3.2&nbsp;&nbsp;</span>Create Submission .CSV File</a></span></li></ul></li></ul></li><li><span><a href="#References" data-toc-modified-id="References-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>References</a></span></li></ul></div>

## Jupyter Notebook Set Up

### Import Libraries

Note: It is assumed that the matplotlib and plotly library has been installed on the user's machine before it is available for import.

```bash
> pip install matplotlib
> pip install plotly
```


#### General Utility Libraries

In [1]:
import os

#### Panda + Numpy

In [2]:
# Python libraries for data manipulation
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import NearestNeighbors

Set up additional libraries for graphical presentations. A number of earlier graphs use Matplotlib. In later sections, Plotly graphs are used.

#### Matplotlib Library

In [3]:
#import matplotlib.pyplot as plt

#### Plotly Libraries

In [4]:
# For interactive graphics and global displays
#import plotly.express as px
#import plotly.graph_objects as go

### <a class="tocSkip">

In [5]:
import warnings
warnings.filterwarnings("ignore")

### <a class="tocSkip">

## Project Data Import, Analysis and Preparation

### Load and Sample Kaggle H&M Datasets

### The assignment starts with a Kaggle H&M Personalised Fashion Recommendations datasets. <a class="tocSkip">

#### Set Up File Paths 

In [6]:
# Set up a Global path file variable to use when loading each dataset
Path_To_Inputs = '../input'

#### H&M Transaction Dataset 

In [7]:
# Load H&M Transaction dataset
df_trxns = pd.read_csv(os.path.join(Path_To_Inputs,'transactions_train.csv'))

In [8]:
# Display Number of Rows, Number of Columns in Transaction dataset prior to Sampling
df_trxns.shape

(31788324, 5)

In [9]:
# Extract random sample of Transactions (without replacement) - 0.1% of records
df_trxns_sample = df_trxns.sample(frac=0.01, replace=False)

In [10]:
# Display Number of Rows, Number of Columns in Transaction dataset after Sampling
df_trxns_sample.shape

(317883, 5)

In [11]:
# Check out the index
df_trxns_sample.head()

t_dat                                        customer_id  \
21121913  2020-01-09  e72a1936c29415d372a8bed6f4d6bf97fc1e0b28d6db4c...   
2723059   2018-11-21  7788e5119c1d72e5d23c0f789b28803a7dd3f9b24a01eb...   
28601425  2020-07-04  b8fcbf8634b17cd88bdf35fa6ff15bcd9a5f52d1999c61...   
5481651   2019-01-27  228b839a5c53bc36a83d96b8cd8c14e92de5f13401ac9d...   
14003024  2019-07-17  9f605930f4b73fa1b6164c47bbc69e3770db7c86b6363b...   

          article_id     price  sales_channel_id  
21121913   805321001  0.025407                 2  
2723059    189626001  0.016932                 2  
28601425   774627006  0.013542                 1  
5481651    686265006  0.025407                 2  
14003024   578311010  0.023475                 2

#### H&M Articles Dataset 

In [12]:
# Read file, located in 'input' folder above this notebook, for articles data - containing meta data on purchasable items.
df_Articles = pd.read_csv(os.path.join(Path_To_Inputs,'articles.csv'))

In [13]:
# Display Number of Rows, Number of Columns in Articles dataset
df_Articles.shape

(105542, 25)

In [14]:
# Extract only the H&M articles contained in the sample Transaction dataset 
articleIDs_in_Trxn_Sample  = set(df_trxns_sample["article_id"])

In [15]:
# Create dataset that only contains H&M articles in the sample Transaction dataset
df_articles_sample = df_Articles[df_Articles["article_id"].isin(articleIDs_in_Trxn_Sample)]

In [16]:
# Display Number of Rows, Number of Columns in Sample Articles dataset
df_articles_sample.shape

(53830, 25)

#### H&M Customers Dataset 

In [17]:
# Load H&M Customer dataset
df_customers = pd.read_csv(os.path.join(Path_To_Inputs,'customers.csv'))

In [18]:
# Display Number of Rows, Number of Columns in Customers dataset prior to Sampling
df_customers.shape

(1371980, 7)

In [19]:
# Extract only the H&M Customers contained in the sample Transaction dataset 
custIDs_in_Trxn_Sample  = set(df_trxns_sample["customer_id"])

In [20]:
#Count Customer Ids in Transaction Sample
len(custIDs_in_Trxn_Sample)

229790

In [21]:
# Create dataset that only contains H&M Customers from the sample Transaction dataset
df_customers_sample = df_customers[df_customers["customer_id"].isin(custIDs_in_Trxn_Sample)]

In [22]:
# Display Number of Rows, Number of Columns in Sample Customers dataset
df_customers_sample.shape

(229790, 7)

###### 

### Exploratory Data Analysis

### Display key data attributes and distributions in Sampled H&M datasets. <a class="tocSkip">

#### H&M Sampled Transaction Dataset 

In [23]:
# Display basic column information for the initial Kaggle 'transaction_train' datatset - after Sampling of original dataset
df_trxns_sample.head()

t_dat                                        customer_id  \
21121913  2020-01-09  e72a1936c29415d372a8bed6f4d6bf97fc1e0b28d6db4c...   
2723059   2018-11-21  7788e5119c1d72e5d23c0f789b28803a7dd3f9b24a01eb...   
28601425  2020-07-04  b8fcbf8634b17cd88bdf35fa6ff15bcd9a5f52d1999c61...   
5481651   2019-01-27  228b839a5c53bc36a83d96b8cd8c14e92de5f13401ac9d...   
14003024  2019-07-17  9f605930f4b73fa1b6164c47bbc69e3770db7c86b6363b...   

          article_id     price  sales_channel_id  
21121913   805321001  0.025407                 2  
2723059    189626001  0.016932                 2  
28601425   774627006  0.013542                 1  
5481651    686265006  0.025407                 2  
14003024   578311010  0.023475                 2

In [24]:
# Display datatype of colums in original H&M Transaction dataset
df_trxns_sample.dtypes

t_dat                object
customer_id          object
article_id            int64
price               float64
sales_channel_id      int64
dtype: object

#### H&M Sampled Customer Dataset 

In [25]:
# Display basic column information for the initial Kaggle 'customers' datatset - after Sampling of original dataset
df_customers_sample.head()

customer_id   FN  Active  \
1   0000423b00ade91418cceaf3b26c6af3dd342b51fd051e...  NaN     NaN   
5   000064249685c11552da43ef22a5030f35a147f723d5b0...  NaN     NaN   
7   00007d2de826758b65a93dd24ce629ed66842531df6699...  1.0     1.0   
13  00009d946eec3ea54add5ba56d5210ea898def4b46c685...  1.0     1.0   
21  0000f1c71aafe5963c3d195cf273f7bfd50bbf17761c91...  1.0     1.0   

   club_member_status fashion_news_frequency   age  \
1              ACTIVE                   NONE  25.0   
5                 NaN                    NaN   NaN   
7              ACTIVE              Regularly  32.0   
13             ACTIVE              Regularly  56.0   
21             ACTIVE              Regularly  30.0   

                                          postal_code  
1   2973abc54daa8a5f8ccfe9362140c63247c5eee03f1d93...  
5   2c29ae653a9282cce4151bd87643c907644e09541abc28...  
7   8d6f45050876d059c830a0fe63f1a4c022de279bb68ce3...  
13  b31984b20a8c478de38eaf113c581ff64e63c4242e607b...  
21  7c902dca60ee0bd0f9030eefd445d11146e8d24835738a...

In [26]:
df_customers_sample.dtypes

customer_id                object
FN                        float64
Active                    float64
club_member_status         object
fashion_news_frequency     object
age                       float64
postal_code                object
dtype: object

#### H&M Sampled Articles Dataset 

In [27]:
# Display basic column information for the initial Kaggle 'article' datatset - after Sampling of original dataset
df_articles_sample.head()

article_id  product_code          prod_name  product_type_no  \
0   108775015        108775          Strap top              253   
1   108775044        108775          Strap top              253   
2   108775051        108775      Strap top (1)              253   
3   110065001        110065  OP T-shirt (Idro)              306   
4   110065002        110065  OP T-shirt (Idro)              306   

  product_type_name  product_group_name  graphical_appearance_no  \
0          Vest top  Garment Upper body                  1010016   
1          Vest top  Garment Upper body                  1010016   
2          Vest top  Garment Upper body                  1010017   
3               Bra           Underwear                  1010016   
4               Bra           Underwear                  1010016   

  graphical_appearance_name  colour_group_code colour_group_name  ...  \
0                     Solid                  9             Black  ...   
1                     Solid                 10             White  ...   
2                    Stripe                 11         Off White  ...   
3                     Solid                  9             Black  ...   
4                     Solid                 10             White  ...   

   department_name index_code        index_name index_group_no  \
0     Jersey Basic          A        Ladieswear              1   
1     Jersey Basic          A        Ladieswear              1   
2     Jersey Basic          A        Ladieswear              1   
3   Clean Lingerie          B  Lingeries/Tights              1   
4   Clean Lingerie          B  Lingeries/Tights              1   

   index_group_name section_no            section_name garment_group_no  \
0        Ladieswear         16  Womens Everyday Basics             1002   
1        Ladieswear         16  Womens Everyday Basics             1002   
2        Ladieswear         16  Womens Everyday Basics             1002   
3        Ladieswear         61         Womens Lingerie             1017   
4        Ladieswear         61         Womens Lingerie             1017   

   garment_group_name                                        detail_desc  
0        Jersey Basic            Jersey top with narrow shoulder straps.  
1        Jersey Basic            Jersey top with narrow shoulder straps.  
2        Jersey Basic            Jersey top with narrow shoulder straps.  
3   Under-, Nightwear  Microfibre T-shirt bra with underwired, moulde...  
4   Under-, Nightwear  Microfibre T-shirt bra with underwired, moulde...  

[5 rows x 25 columns]

In [28]:
df_articles_sample.dtypes

article_id                       int64
product_code                     int64
prod_name                       object
product_type_no                  int64
product_type_name               object
product_group_name              object
graphical_appearance_no          int64
graphical_appearance_name       object
colour_group_code                int64
colour_group_name               object
perceived_colour_value_id        int64
perceived_colour_value_name     object
perceived_colour_master_id       int64
perceived_colour_master_name    object
department_no                    int64
department_name                 object
index_code                      object
index_name                      object
index_group_no                   int64
index_group_name                object
section_no                       int64
section_name                    object
garment_group_no                 int64
garment_group_name              object
detail_desc                     object
dtype: object

###### 

### Data Preparation and Feature Enrichment

### Construct a dataset upon which to build ML models. <a class="tocSkip">

### Begin with Transaction Dataset (sampled) and merge in customer and article data. <a class="tocSkip">

#### Transaction Data Feature Enrichment (Sample)

In [29]:
# Create new dataset for modelling. Initial basis for new dataset is the sampled transaction dataset.
df_HM_CustTrxn = df_trxns_sample

In [30]:
df_HM_CustTrxn.head()

t_dat                                        customer_id  \
21121913  2020-01-09  e72a1936c29415d372a8bed6f4d6bf97fc1e0b28d6db4c...   
2723059   2018-11-21  7788e5119c1d72e5d23c0f789b28803a7dd3f9b24a01eb...   
28601425  2020-07-04  b8fcbf8634b17cd88bdf35fa6ff15bcd9a5f52d1999c61...   
5481651   2019-01-27  228b839a5c53bc36a83d96b8cd8c14e92de5f13401ac9d...   
14003024  2019-07-17  9f605930f4b73fa1b6164c47bbc69e3770db7c86b6363b...   

          article_id     price  sales_channel_id  
21121913   805321001  0.025407                 2  
2723059    189626001  0.016932                 2  
28601425   774627006  0.013542                 1  
5481651    686265006  0.025407                 2  
14003024   578311010  0.023475                 2

In [31]:
# Decompose Datetime information 
df_HM_CustTrxn['t_dat'] = pd.to_datetime(df_HM_CustTrxn['t_dat'], format="%Y-%m-%d")

# Add Column for Transaction Year and convert to integer value
df_HM_CustTrxn['txn_year'] = df_HM_CustTrxn['t_dat'].dt.strftime('%Y')
df_HM_CustTrxn['txn_year'] = df_HM_CustTrxn['txn_year'].astype(str).astype(int)

# Add Column for Transaction Month and convert to integer value
df_HM_CustTrxn['txn_mth'] = df_HM_CustTrxn['t_dat'].dt.strftime('%m')
df_HM_CustTrxn['txn_mth'] = df_HM_CustTrxn['txn_mth'].astype(str).astype(int)

# Add Column for Transaction Day and convert to integer value
df_HM_CustTrxn['txn_day'] = df_HM_CustTrxn['t_dat'].dt.strftime('%d')
df_HM_CustTrxn['txn_day'] = df_HM_CustTrxn['txn_day'].astype(str).astype(int)

In [32]:
# Filter out column to drop timestamp attribute
# df_HM_CustTrxn = df_HM_CustTrxn.iloc[:,1:8]

In [33]:
df_HM_CustTrxn.head()

t_dat                                        customer_id  \
21121913 2020-01-09  e72a1936c29415d372a8bed6f4d6bf97fc1e0b28d6db4c...   
2723059  2018-11-21  7788e5119c1d72e5d23c0f789b28803a7dd3f9b24a01eb...   
28601425 2020-07-04  b8fcbf8634b17cd88bdf35fa6ff15bcd9a5f52d1999c61...   
5481651  2019-01-27  228b839a5c53bc36a83d96b8cd8c14e92de5f13401ac9d...   
14003024 2019-07-17  9f605930f4b73fa1b6164c47bbc69e3770db7c86b6363b...   

          article_id     price  sales_channel_id  txn_year  txn_mth  txn_day  
21121913   805321001  0.025407                 2      2020        1        9  
2723059    189626001  0.016932                 2      2018       11       21  
28601425   774627006  0.013542                 1      2020        7        4  
5481651    686265006  0.025407                 2      2019        1       27  
14003024   578311010  0.023475                 2      2019        7       17

In [34]:
# Check for null values in the Transaction Dataframe
[print('Any Null Values in the Trxn (sample) dataframe?','\n' ,df_HM_CustTrxn.isnull().any())]

Any Null Values in the Trxn (sample) dataframe? 
 t_dat               False
customer_id         False
article_id          False
price               False
sales_channel_id    False
txn_year            False
txn_mth             False
txn_day             False
dtype: bool


[None]

#### Extend Dataframe with Customer Data

In [35]:
# Confirm dimensions of Trxn (sample) dataframe BEFORE merging with Customer data
df_HM_CustTrxn.shape

(317883, 8)

#### Merge Customer Attributes into the Trxn (Sample) Dataframe. <a class="tocSkip">

In [36]:
# Merge Customer atributes from the Customer dataset based on a join with the 'customer_id' field
df_HM_CustTrxn = pd.merge(df_HM_CustTrxn, df_customers_sample, how="left", on=["customer_id"])

In [37]:
# Confirm dimensions of Trxn (sample) dataframe AFTER merging with Customer data
df_HM_CustTrxn.shape

(317883, 14)

In [38]:
df_HM_CustTrxn.head()

t_dat                                        customer_id  article_id  \
0 2020-01-09  e72a1936c29415d372a8bed6f4d6bf97fc1e0b28d6db4c...   805321001   
1 2018-11-21  7788e5119c1d72e5d23c0f789b28803a7dd3f9b24a01eb...   189626001   
2 2020-07-04  b8fcbf8634b17cd88bdf35fa6ff15bcd9a5f52d1999c61...   774627006   
3 2019-01-27  228b839a5c53bc36a83d96b8cd8c14e92de5f13401ac9d...   686265006   
4 2019-07-17  9f605930f4b73fa1b6164c47bbc69e3770db7c86b6363b...   578311010   

      price  sales_channel_id  txn_year  txn_mth  txn_day   FN  Active  \
0  0.025407                 2      2020        1        9  1.0     1.0   
1  0.016932                 2      2018       11       21  1.0     1.0   
2  0.013542                 1      2020        7        4  NaN     NaN   
3  0.025407                 2      2019        1       27  NaN     NaN   
4  0.023475                 2      2019        7       17  1.0     1.0   

  club_member_status fashion_news_frequency   age  \
0             ACTIVE              Regularly  26.0   
1             ACTIVE              Regularly  24.0   
2             ACTIVE                   NONE  50.0   
3             ACTIVE                   NONE  48.0   
4             ACTIVE              Regularly  25.0   

                                         postal_code  
0  20ae49d1cda483a61a076ad91cabb23a603c7192850b84...  
1  16e03025d65e2909d4fe29d265b6138945bb1c274ece6a...  
2  6f60be2a66d802f01ac9b92b22366cd1a74b76aa6a1aeb...  
3  bd4b2920f04833514aa85e07ea1d12cb41bc7794a7100b...  
4  231f92f3036f2b12e54e9c60446da74f7cb58e7f509ce8...

In [39]:
# Filter out other Customer specific attributes apart from; age + fashion_news_frequency + club_member_status + postal code
# There are very many missing values in the other Customer dataset columns, so these will be ignored
#df_HM_CustTrxn = df_HM_CustTrxn.iloc[:, [0, 1, 2, 3, 4, 5, 6, 9, 10, 11, 12]]
df_HM_CustTrxn.drop(["FN", "Active"], axis=1, inplace=True)

In [40]:
df_HM_CustTrxn.head()

t_dat                                        customer_id  article_id  \
0 2020-01-09  e72a1936c29415d372a8bed6f4d6bf97fc1e0b28d6db4c...   805321001   
1 2018-11-21  7788e5119c1d72e5d23c0f789b28803a7dd3f9b24a01eb...   189626001   
2 2020-07-04  b8fcbf8634b17cd88bdf35fa6ff15bcd9a5f52d1999c61...   774627006   
3 2019-01-27  228b839a5c53bc36a83d96b8cd8c14e92de5f13401ac9d...   686265006   
4 2019-07-17  9f605930f4b73fa1b6164c47bbc69e3770db7c86b6363b...   578311010   

      price  sales_channel_id  txn_year  txn_mth  txn_day club_member_status  \
0  0.025407                 2      2020        1        9             ACTIVE   
1  0.016932                 2      2018       11       21             ACTIVE   
2  0.013542                 1      2020        7        4             ACTIVE   
3  0.025407                 2      2019        1       27             ACTIVE   
4  0.023475                 2      2019        7       17             ACTIVE   

  fashion_news_frequency   age  \
0              Regularly  26.0   
1              Regularly  24.0   
2                   NONE  50.0   
3                   NONE  48.0   
4              Regularly  25.0   

                                         postal_code  
0  20ae49d1cda483a61a076ad91cabb23a603c7192850b84...  
1  16e03025d65e2909d4fe29d265b6138945bb1c274ece6a...  
2  6f60be2a66d802f01ac9b92b22366cd1a74b76aa6a1aeb...  
3  bd4b2920f04833514aa85e07ea1d12cb41bc7794a7100b...  
4  231f92f3036f2b12e54e9c60446da74f7cb58e7f509ce8...

In [41]:
df_HM_CustTrxn.shape

(317883, 12)

In [42]:
# Check for null values in the merged Trxn/Customer dataframe
[print('Any Null Values in the merged Trxn/Customer dataframe?','\n' ,df_HM_CustTrxn.isnull().any())]

Any Null Values in the merged Trxn/Customer dataframe? 
 t_dat                     False
customer_id               False
article_id                False
price                     False
sales_channel_id          False
txn_year                  False
txn_mth                   False
txn_day                   False
club_member_status         True
fashion_news_frequency     True
age                        True
postal_code               False
dtype: bool


[None]

#### Remove Rows with Missing Club Member Status Attributes. <a class="tocSkip">

In [43]:
# Check missing values in the club_member_status column  
df_null_club_member_status = df_HM_CustTrxn.loc[(df_HM_CustTrxn["club_member_status"].isnull())]

[print('Numbers of rows with missing club member status?','\n' ,len(df_null_club_member_status.index))]
[print('\nPrecentage of total rows in Trxn dataframe with missing club member status attributes?\n{:06.3f}%'
                               .format((len(df_null_club_member_status.index)/len(df_HM_CustTrxn.index))*100))]


# For simplicity the rows with missing values in the club member status will be ignored
# Re-Build Txn/Customer Dataframe with no missing values in the club member status
df_HM_CustTrxn = df_HM_CustTrxn.loc[(df_HM_CustTrxn["club_member_status"].notnull())]

# Confirm dimensions of Trxn (sample) dataframe AFTER removing rows with missing values in the club member status column
[print('\nCurrent Number of Rows/Columns in Trxn/Customer dataframe - with missing club mem status values removed','\n' ,
                                                                                                   df_HM_CustTrxn.shape)]


Numbers of rows with missing club member status? 
 602

Precentage of total rows in Trxn dataframe with missing club member status attributes?
00.189%

Current Number of Rows/Columns in Trxn/Customer dataframe - with missing club mem status values removed 
 (317281, 12)


[None]

#### Remove Rows with Missing Fashion News Frequency Attributes. <a class="tocSkip">

In [44]:
# Check missing values in the club_member_status column  
df_null_fn_freq = df_HM_CustTrxn.loc[(df_HM_CustTrxn["fashion_news_frequency"].isnull())]

[print('Numbers of rows with missing fashion_news_frequency values?','\n' ,len(df_null_fn_freq.index))]
[print('\nPrecentage of total rows in Trxn dataframe with missing fashion_news_frequency attributes?\n{:06.3f}%'
                               .format((len(df_null_fn_freq.index)/len(df_HM_CustTrxn.index))*100))]

# For simplicity the rows with missing values in the fashion_news_frequency column will be ignored
# Re-Build Txn/Customer Dataframe with no missing values in the fashion_news_frequency column
df_HM_CustTrxn = df_HM_CustTrxn.loc[(df_HM_CustTrxn["fashion_news_frequency"].notnull())]

# Confirm dimensions of Trxn (sample) dataframe AFTER removing rows with missing values in the fashion_news_frequency column
[print('\nCurrent Number of Rows/Columns in Trxn/Customer dataframe - missing Fash News Freq values removed','\n' ,
                                                                                                   df_HM_CustTrxn.shape)]

Numbers of rows with missing fashion_news_frequency values? 
 1320

Precentage of total rows in Trxn dataframe with missing fashion_news_frequency attributes?
00.416%

Current Number of Rows/Columns in Trxn/Customer dataframe - missing Fash News Freq values removed 
 (315961, 12)


[None]

#### Remove Rows with Missing Age Attributes. <a class="tocSkip">

In [45]:
# Check missing values in the age column  
df_null_age = df_HM_CustTrxn.loc[(df_HM_CustTrxn["age"].isnull())]

[print('Numbers of rows with missing age values?','\n' ,len(df_null_age.index))]
[print('\nPrecentage of total rows in Trxn dataframe with missing age attributes?\n{:06.3f}%'
                               .format((len(df_null_age.index)/len(df_HM_CustTrxn.index))*100))]

# Re-Build Txn/Customer Dataframe with no missing values in the age column
df_HM_CustTrxn = df_HM_CustTrxn.loc[(df_HM_CustTrxn["age"].notnull())]


# Confirm dimensions of Trxn (sample) dataframe AFTER removing rows with missing values in the age column
[print('\nCurrent Number of Rows/Columns in Trxn/Customer dataframe - missing ages removed','\n' ,df_HM_CustTrxn.shape)]

Numbers of rows with missing age values? 
 1271

Precentage of total rows in Trxn dataframe with missing age attributes?
00.402%

Current Number of Rows/Columns in Trxn/Customer dataframe - missing ages removed 
 (314690, 12)


[None]

#### Encode the club_member_status + fashion_news_frequency attributes <a class="tocSkip">

In [46]:
df_HM_CustTrxn.head()

t_dat                                        customer_id  article_id  \
0 2020-01-09  e72a1936c29415d372a8bed6f4d6bf97fc1e0b28d6db4c...   805321001   
1 2018-11-21  7788e5119c1d72e5d23c0f789b28803a7dd3f9b24a01eb...   189626001   
2 2020-07-04  b8fcbf8634b17cd88bdf35fa6ff15bcd9a5f52d1999c61...   774627006   
3 2019-01-27  228b839a5c53bc36a83d96b8cd8c14e92de5f13401ac9d...   686265006   
4 2019-07-17  9f605930f4b73fa1b6164c47bbc69e3770db7c86b6363b...   578311010   

      price  sales_channel_id  txn_year  txn_mth  txn_day club_member_status  \
0  0.025407                 2      2020        1        9             ACTIVE   
1  0.016932                 2      2018       11       21             ACTIVE   
2  0.013542                 1      2020        7        4             ACTIVE   
3  0.025407                 2      2019        1       27             ACTIVE   
4  0.023475                 2      2019        7       17             ACTIVE   

  fashion_news_frequency   age  \
0              Regularly  26.0   
1              Regularly  24.0   
2                   NONE  50.0   
3                   NONE  48.0   
4              Regularly  25.0   

                                         postal_code  
0  20ae49d1cda483a61a076ad91cabb23a603c7192850b84...  
1  16e03025d65e2909d4fe29d265b6138945bb1c274ece6a...  
2  6f60be2a66d802f01ac9b92b22366cd1a74b76aa6a1aeb...  
3  bd4b2920f04833514aa85e07ea1d12cb41bc7794a7100b...  
4  231f92f3036f2b12e54e9c60446da74f7cb58e7f509ce8...

In [47]:
def encode_Categor_Features(df, col_name):
       
    # The categorical values in dataframe must be encoded to numberical values
    df_encoded_cols = pd.get_dummies(df[col_name])
    df_encoded      = df.join(df_encoded_cols)
    
    # Remove the prior Categorical column from the Articles dataframe
    df_encoded.drop([col_name], axis=1, inplace=True)
    
    # Return the dataframe with encoded columsn added
    return df_encoded

In [48]:
cust_cols_to_encode = ["club_member_status","fashion_news_frequency"]
for column_name in cust_cols_to_encode:
    df_HM_CustTrxn = encode_Categor_Features(df_HM_CustTrxn, column_name)

In [49]:
# Drop Postal Code - a non numerical value that is not easily converted
df_HM_CustTrxn.drop(["postal_code"], axis=1, inplace=True)

In [50]:
# Display basic column information for the initial Kaggle 'article' datatset - after Sampling of original dataset
df_HM_CustTrxn.head()

t_dat                                        customer_id  article_id  \
0 2020-01-09  e72a1936c29415d372a8bed6f4d6bf97fc1e0b28d6db4c...   805321001   
1 2018-11-21  7788e5119c1d72e5d23c0f789b28803a7dd3f9b24a01eb...   189626001   
2 2020-07-04  b8fcbf8634b17cd88bdf35fa6ff15bcd9a5f52d1999c61...   774627006   
3 2019-01-27  228b839a5c53bc36a83d96b8cd8c14e92de5f13401ac9d...   686265006   
4 2019-07-17  9f605930f4b73fa1b6164c47bbc69e3770db7c86b6363b...   578311010   

      price  sales_channel_id  txn_year  txn_mth  txn_day   age  ACTIVE  \
0  0.025407                 2      2020        1        9  26.0       1   
1  0.016932                 2      2018       11       21  24.0       1   
2  0.013542                 1      2020        7        4  50.0       1   
3  0.025407                 2      2019        1       27  48.0       1   
4  0.023475                 2      2019        7       17  25.0       1   

   LEFT CLUB  PRE-CREATE  Monthly  NONE  Regularly  
0          0           0        0     0          1  
1          0           0        0     0          1  
2          0           0        0     1          0  
3          0           0        0     1          0  
4          0           0        0     0          1

In [51]:
# Confirm dimensions of Trxn (sample) dataframe AFTER encoding Categorical Values
[print('\nCurrent Number of Rows/Columns in Trxn/Customer dataframe - after encoding of categorical values completed','\n' ,
                                                                                                   df_HM_CustTrxn.shape)]


Current Number of Rows/Columns in Trxn/Customer dataframe - after encoding of categorical values completed 
 (314690, 15)


[None]

In [52]:
# Check for null values in the Transaction Dataframe
[print('Any Null Values in the merged Trxn/Customer (sample) dataframe?','\n' ,df_HM_CustTrxn.isnull().any())]

Any Null Values in the merged Trxn/Customer (sample) dataframe? 
 t_dat               False
customer_id         False
article_id          False
price               False
sales_channel_id    False
txn_year            False
txn_mth             False
txn_day             False
age                 False
ACTIVE              False
LEFT CLUB           False
PRE-CREATE          False
Monthly             False
NONE                False
Regularly           False
dtype: bool


[None]

#### Extend Dataframe with Articles Data

#### Remove 'Redundant' Categorical Columns in Articles (sample) Dataframe. <a class="tocSkip">

In [53]:
# Display basic column information for the initial Kaggle 'article' datatset - after Sampling of original dataset
df_articles_sample.head()

article_id  product_code          prod_name  product_type_no  \
0   108775015        108775          Strap top              253   
1   108775044        108775          Strap top              253   
2   108775051        108775      Strap top (1)              253   
3   110065001        110065  OP T-shirt (Idro)              306   
4   110065002        110065  OP T-shirt (Idro)              306   

  product_type_name  product_group_name  graphical_appearance_no  \
0          Vest top  Garment Upper body                  1010016   
1          Vest top  Garment Upper body                  1010016   
2          Vest top  Garment Upper body                  1010017   
3               Bra           Underwear                  1010016   
4               Bra           Underwear                  1010016   

  graphical_appearance_name  colour_group_code colour_group_name  ...  \
0                     Solid                  9             Black  ...   
1                     Solid                 10             White  ...   
2                    Stripe                 11         Off White  ...   
3                     Solid                  9             Black  ...   
4                     Solid                 10             White  ...   

   department_name index_code        index_name index_group_no  \
0     Jersey Basic          A        Ladieswear              1   
1     Jersey Basic          A        Ladieswear              1   
2     Jersey Basic          A        Ladieswear              1   
3   Clean Lingerie          B  Lingeries/Tights              1   
4   Clean Lingerie          B  Lingeries/Tights              1   

   index_group_name section_no            section_name garment_group_no  \
0        Ladieswear         16  Womens Everyday Basics             1002   
1        Ladieswear         16  Womens Everyday Basics             1002   
2        Ladieswear         16  Womens Everyday Basics             1002   
3        Ladieswear         61         Womens Lingerie             1017   
4        Ladieswear         61         Womens Lingerie             1017   

   garment_group_name                                        detail_desc  
0        Jersey Basic            Jersey top with narrow shoulder straps.  
1        Jersey Basic            Jersey top with narrow shoulder straps.  
2        Jersey Basic            Jersey top with narrow shoulder straps.  
3   Under-, Nightwear  Microfibre T-shirt bra with underwired, moulde...  
4   Under-, Nightwear  Microfibre T-shirt bra with underwired, moulde...  

[5 rows x 25 columns]

In [54]:
# Remove 'redundant' categotical columns from Articles (sample) dataframe
# Consider a categorical column redundant when is 'partnered' with a numerical code 

redundo_catcols_to_drop =  ["prod_name", 
                            "product_type_name", 
                            "graphical_appearance_name",
                            "colour_group_name",
                            "perceived_colour_value_name",
                            "perceived_colour_master_name",
                            "department_name",
                            "index_name",
                            "index_group_name",
                            "section_name",
                            "garment_group_name",
                            "detail_desc"]


df_articles_sample.drop(redundo_catcols_to_drop, axis=1, inplace=True)

In [55]:
# Display basic column information for the initial Kaggle 'article' datatset - after Sampling of original dataset
df_articles_sample.head()

article_id  product_code  product_type_no  product_group_name  \
0   108775015        108775              253  Garment Upper body   
1   108775044        108775              253  Garment Upper body   
2   108775051        108775              253  Garment Upper body   
3   110065001        110065              306           Underwear   
4   110065002        110065              306           Underwear   

   graphical_appearance_no  colour_group_code  perceived_colour_value_id  \
0                  1010016                  9                          4   
1                  1010016                 10                          3   
2                  1010017                 11                          1   
3                  1010016                  9                          4   
4                  1010016                 10                          3   

   perceived_colour_master_id  department_no index_code  index_group_no  \
0                           5           1676          A               1   
1                           9           1676          A               1   
2                           9           1676          A               1   
3                           5           1339          B               1   
4                           9           1339          B               1   

   section_no  garment_group_no  
0          16              1002  
1          16              1002  
2          16              1002  
3          61              1017  
4          61              1017

#### Use One-Hot-Encoding on remaining categorical columns. <a class="tocSkip">

In [56]:
article_cols_to_encode =['product_group_name', 'index_code']
for column_name in article_cols_to_encode:
    df_articles_sample = encode_Categor_Features(df_articles_sample, column_name)

In [57]:
df_articles_sample.shape

(53830, 39)

In [58]:
# Display basic column information for sampled articles dataframe after encoding
df_articles_sample.head()

article_id  product_code  product_type_no  graphical_appearance_no  \
0   108775015        108775              253                  1010016   
1   108775044        108775              253                  1010016   
2   108775051        108775              253                  1010017   
3   110065001        110065              306                  1010016   
4   110065002        110065              306                  1010016   

   colour_group_code  perceived_colour_value_id  perceived_colour_master_id  \
0                  9                          4                           5   
1                 10                          3                           9   
2                 11                          1                           9   
3                  9                          4                           5   
4                 10                          3                           9   

   department_no  index_group_no  section_no  ...  A  B  C  D  F  G  H  I  J  \
0           1676               1          16  ...  1  0  0  0  0  0  0  0  0   
1           1676               1          16  ...  1  0  0  0  0  0  0  0  0   
2           1676               1          16  ...  1  0  0  0  0  0  0  0  0   
3           1339               1          61  ...  0  1  0  0  0  0  0  0  0   
4           1339               1          61  ...  0  1  0  0  0  0  0  0  0   

   S  
0  0  
1  0  
2  0  
3  0  
4  0  

[5 rows x 39 columns]

#### Merge Trxn/Customers Dataframe with Articles Dataframe (sample). <a class="tocSkip">

In [59]:
# Merge Articles atributes from the Articles dataset based on a join with the 'article_id' field
df_HM_CustTrxnArtcls = pd.merge(df_HM_CustTrxn, df_articles_sample, how="left", on=["article_id"])

In [60]:
# Confirm dimensions of Trxn (sample) dataframe AFTER merging with Articles data
df_HM_CustTrxnArtcls.shape

(314690, 53)

In [61]:
# Display basic column information for the initial Kaggle 'transaction_train' datatset - AFTER merging with Articles data
df_HM_CustTrxnArtcls.head()

t_dat                                        customer_id  article_id  \
0 2020-01-09  e72a1936c29415d372a8bed6f4d6bf97fc1e0b28d6db4c...   805321001   
1 2018-11-21  7788e5119c1d72e5d23c0f789b28803a7dd3f9b24a01eb...   189626001   
2 2020-07-04  b8fcbf8634b17cd88bdf35fa6ff15bcd9a5f52d1999c61...   774627006   
3 2019-01-27  228b839a5c53bc36a83d96b8cd8c14e92de5f13401ac9d...   686265006   
4 2019-07-17  9f605930f4b73fa1b6164c47bbc69e3770db7c86b6363b...   578311010   

      price  sales_channel_id  txn_year  txn_mth  txn_day   age  ACTIVE  ...  \
0  0.025407                 2      2020        1        9  26.0       1  ...   
1  0.016932                 2      2018       11       21  24.0       1  ...   
2  0.013542                 1      2020        7        4  50.0       1  ...   
3  0.025407                 2      2019        1       27  48.0       1  ...   
4  0.023475                 2      2019        7       17  25.0       1  ...   

   A  B  C  D  F  G  H  I  J  S  
0  1  0  0  0  0  0  0  0  0  0  
1  0  0  0  1  0  0  0  0  0  0  
2  0  0  0  0  1  0  0  0  0  0  
3  0  0  0  0  0  0  1  0  0  0  
4  0  0  1  0  0  0  0  0  0  0  

[5 rows x 53 columns]

In [62]:
# Check for null values in the merged Trxn/Customer/Articles dataframe
[print('Any Null Values in the merged Trxn/Customer/Articles dataframe?','\n' ,df_HM_CustTrxnArtcls.isnull().any())]

Any Null Values in the merged Trxn/Customer/Articles dataframe? 
 t_dat                         False
customer_id                   False
article_id                    False
price                         False
sales_channel_id              False
txn_year                      False
txn_mth                       False
txn_day                       False
age                           False
ACTIVE                        False
LEFT CLUB                     False
PRE-CREATE                    False
Monthly                       False
NONE                          False
Regularly                     False
product_code                  False
product_type_no               False
graphical_appearance_no       False
colour_group_code             False
perceived_colour_value_id     False
perceived_colour_master_id    False
department_no                 False
index_group_no                False
section_no                    False
garment_group_no              False
Accessories                   Fals

[None]

#### Final Data Preparation Steps

#### Remove Any Duplicates in Final Merged Dataframe. <a class="tocSkip">

In [63]:
# run a check to see if any duplicate entires exist in the merged Trxn/Customer/Articles dataframe
[print('Numbers of duplicated rows in the final merged Trxn/Customer/Articles dataframe?','\n' 
                                                                       ,df_HM_CustTrxnArtcls.duplicated().sum())]

Numbers of duplicated rows in the final merged Trxn/Customer/Articles dataframe? 
 400


[None]

In [64]:
# Drop duplicate rows in the merged Trxn/Customer/Articles dataframe
df_HM_CustTrxnArtcls = df_HM_CustTrxnArtcls.drop_duplicates()

In [65]:
# Re-check: Find duplicate rows across all columns
duplicateRows = df_HM_CustTrxnArtcls[df_HM_CustTrxnArtcls.duplicated()]

In [66]:
# This code will advise if duplicates are successfully removed.
if duplicateRows.shape[0] < 1:
    [print('Success! Duplicate Rows Removed from merged Trxn/Customer/Articles dataframe')]
else:
    [print('Warning! Duplicate Rows Remain in merged Trxn/Customer/Articles dataframe')] 

Success! Duplicate Rows Removed from merged Trxn/Customer/Articles dataframe


In [67]:
# Confirm dimensions of Trxn/Customer/Articles (sample) dataframe AFTER removing duplicate rows
[print('\nCurrent Number of Rows/Columns in merged Trxn/Customer/Articles dataframe - duplicates removed','\n' ,
                                                                                           df_HM_CustTrxnArtcls.shape)]


Current Number of Rows/Columns in merged Trxn/Customer/Articles dataframe - duplicates removed 
 (314290, 53)


[None]

In [68]:
### See if there is a way to check the integrity of the customer merging..

In [69]:
df_HM_CustTrxnArtcls.dtypes

t_dat                         datetime64[ns]
customer_id                           object
article_id                             int64
price                                float64
sales_channel_id                       int64
txn_year                               int32
txn_mth                                int32
txn_day                                int32
age                                  float64
ACTIVE                                 uint8
LEFT CLUB                              uint8
PRE-CREATE                             uint8
Monthly                                uint8
NONE                                   uint8
Regularly                              uint8
product_code                           int64
product_type_no                        int64
graphical_appearance_no                int64
colour_group_code                      int64
perceived_colour_value_id              int64
perceived_colour_master_id             int64
department_no                          int64
index_grou

###### 

## Split Data: Train and Test Sets

In [70]:
df_HM_CustTrxnArtcls.head()

t_dat                                        customer_id  article_id  \
0 2020-01-09  e72a1936c29415d372a8bed6f4d6bf97fc1e0b28d6db4c...   805321001   
1 2018-11-21  7788e5119c1d72e5d23c0f789b28803a7dd3f9b24a01eb...   189626001   
2 2020-07-04  b8fcbf8634b17cd88bdf35fa6ff15bcd9a5f52d1999c61...   774627006   
3 2019-01-27  228b839a5c53bc36a83d96b8cd8c14e92de5f13401ac9d...   686265006   
4 2019-07-17  9f605930f4b73fa1b6164c47bbc69e3770db7c86b6363b...   578311010   

      price  sales_channel_id  txn_year  txn_mth  txn_day   age  ACTIVE  ...  \
0  0.025407                 2      2020        1        9  26.0       1  ...   
1  0.016932                 2      2018       11       21  24.0       1  ...   
2  0.013542                 1      2020        7        4  50.0       1  ...   
3  0.025407                 2      2019        1       27  48.0       1  ...   
4  0.023475                 2      2019        7       17  25.0       1  ...   

   A  B  C  D  F  G  H  I  J  S  
0  1  0  0  0  0  0  0  0  0  0  
1  0  0  0  1  0  0  0  0  0  0  
2  0  0  0  0  1  0  0  0  0  0  
3  0  0  0  0  0  0  1  0  0  0  
4  0  0  1  0  0  0  0  0  0  0  

[5 rows x 53 columns]

In [71]:
df_HM_CustTrxnArtcls.dtypes

t_dat                         datetime64[ns]
customer_id                           object
article_id                             int64
price                                float64
sales_channel_id                       int64
txn_year                               int32
txn_mth                                int32
txn_day                                int32
age                                  float64
ACTIVE                                 uint8
LEFT CLUB                              uint8
PRE-CREATE                             uint8
Monthly                                uint8
NONE                                   uint8
Regularly                              uint8
product_code                           int64
product_type_no                        int64
graphical_appearance_no                int64
colour_group_code                      int64
perceived_colour_value_id              int64
perceived_colour_master_id             int64
department_no                          int64
index_grou

In [72]:
from sklearn.preprocessing import MinMaxScaler
# Normalizing numerical features so that each feature has a value between 0 and 1
scaler = MinMaxScaler()

In [73]:
def scale_trainset_features(df, dtype, no_scale):
    for x in df.columns[df.dtypes == dtype]:
        if x != no_scale:
            df[x] = scaler.fit_transform(np.array(df[x]).reshape(-1,1))    

In [74]:
no_scale = 'article_id'
scale_dtypes = ['int64','int32','float64']
for data_types in scale_dtypes:
    scale_trainset_features(df_HM_CustTrxnArtcls, data_types, no_scale)

#### Split Data into 70% for Training, 30% for Testing. <a class="tocSkip">

In [75]:
# Split the data into train and test set
# Parameter set to 30% for size of Test set
# Given that there is ordering in the dataset, the 'shuffle' parameter is set to 'True'
train, test = train_test_split(df_HM_CustTrxnArtcls, test_size=0.3, random_state=101, shuffle=True)

In [76]:
# Output Confirmation of data split
dataDescription = 'H&M Sample Data Split'
print("\n{} Training Set Shape : \n\t".format(dataDescription))
print(train.shape)
print("\n{} Test Set Shape : \n\t".format(dataDescription))
print(test.shape)


H&M Sample Data Split Training Set Shape : 
	
(220003, 53)

H&M Sample Data Split Test Set Shape : 
	
(94287, 53)


In [77]:
train.head()

t_dat                                        customer_id  \
134705 2020-07-16  bc278ea87e8891c9337a012fce80c0c49b660eb84bfbcc...   
255931 2019-06-04  1751df5330b9d5a270c936232e524ddbc8f7e3192c08ae...   
10837  2020-02-18  2574f05cc73eb0cd2d9e813c2008363e0424dfa5ca743c...   
272598 2019-04-24  d36834d69219148fe281b4c9d6ffbb977668b1bcc488e6...   
153475 2019-04-13  511ae31df843efe116ef54945214731c3165aad2f22190...   

        article_id     price  sales_channel_id  txn_year   txn_mth   txn_day  \
134705   850176003  0.059947               1.0       1.0  0.545455  0.500000   
255931   513373010  0.036530               1.0       0.5  0.454545  0.100000   
10837    759482001  0.033185               1.0       1.0  0.090909  0.566667   
272598   739819002  0.066638               1.0       0.5  0.272727  0.766667   
153475   673930003  0.033185               1.0       0.5  0.272727  0.400000   

             age  ACTIVE  ...  A  B  C  D  F  G  H  I  J  S  
134705  0.421687       1  ...  0  1  0  0  0  0  0  0  0  0  
255931  0.433735       1  ...  1  0  0  0  0  0  0  0  0  0  
10837   0.289157       1  ...  0  0  0  1  0  0  0  0  0  0  
272598  0.530120       1  ...  1  0  0  0  0  0  0  0  0  0  
153475  0.144578       1  ...  0  0  0  1  0  0  0  0  0  0  

[5 rows x 53 columns]

In [78]:
test.head()

t_dat                                        customer_id  \
72997  2018-11-26  17a4fbba8d8b728947c3d966e134ceaa5248be4bf8b88b...   
274139 2020-04-11  e1b69905e60da797f8d92946aa66e95fe6b46dedbd8bcf...   
271987 2018-12-19  ea82914cbb33f658f1cd76af29e7fe048f988302201be1...   
66429  2020-04-14  b64af996030aaa96c2e0b8228150decdf4623c90d6598e...   
132823 2018-10-24  90ae758dbc94409920a2823f14674fa430a08d8faa76bf...   

        article_id     price  sales_channel_id  txn_year   txn_mth   txn_day  \
72997    448515027  0.049911               1.0       0.0  0.909091  0.833333   
274139   803070001  0.033553               1.0       1.0  0.272727  0.333333   
271987   574750002  0.046566               1.0       0.0  1.000000  0.600000   
66429    547780031  0.049911               1.0       1.0  0.272727  0.433333   
132823   678073001  0.049911               1.0       0.0  0.818182  0.766667   

             age  ACTIVE  ...  A  B  C  D  F  G  H  I  J  S  
72997   0.072289       1  ...  0  0  0  1  0  0  0  0  0  0  
274139  0.108434       1  ...  0  1  0  0  0  0  0  0  0  0  
271987  0.469880       1  ...  0  0  1  0  0  0  0  0  0  0  
66429   0.156627       1  ...  0  0  0  1  0  0  0  0  0  0  
132823  0.409639       1  ...  0  0  0  1  0  0  0  0  0  0  

[5 rows x 53 columns]

In [79]:
custs = test.customer_id.unique()

In [80]:
custs.shape

(83722,)

In [81]:
custs

array(['17a4fbba8d8b728947c3d966e134ceaa5248be4bf8b88b06c5d98695c76f5fed',
       'e1b69905e60da797f8d92946aa66e95fe6b46dedbd8bcf807d22e12f091152a8',
       'ea82914cbb33f658f1cd76af29e7fe048f988302201be1b7bcab0655f1b45781',
       ...,
       'afbd69283b575d286b530b833186fbcd077a9ae1fb1102dc9616899e7f14918e',
       '6f44a1ff24ed05d3c3576e70be705db0e53e81f7c7c33071a86e68a10af075ca',
       '5b3e1d50927dbd66de584f9831b69767d1d8f75356fd81ededf1eaae34e00870'],
      dtype=object)

###### 

## Model Preparation

### Set Up the Scaling Algorithm

In [82]:
from sklearn.preprocessing import MinMaxScaler
# Normalizing numerical features so that each feature has a value between 0 and 1
#scaler = MinMaxScaler()

In [83]:
# Normalizing numerical features so that each feature has mean 0 and variance 1
# feature_scaler = StandardScaler()

### Set Up the Columns to be Scaled

In [84]:
nn_cols_old = ['price',
              'age',
              'product_code',
              'product_type_no',
              'graphical_appearance_no',
              'colour_group_code',
              'perceived_colour_value_id',
              'perceived_colour_master_id',
              'department_no',
              'index_group_no',
              'section_no',
              'garment_group_no']

In [85]:
nn_cols =    ['price',
              'age',
              'product_code',
              'product_type_no',
              'graphical_appearance_no',
              'colour_group_code',
              'perceived_colour_value_id',
              'perceived_colour_master_id',
              'department_no',
              'index_group_no',
              'section_no',
              'garment_group_no',
              'Accessories',
              'Bags',
              'Cosmetic',
              'Furniture',
              'Garment Full body',
              'Garment Lower body',
              'Garment Upper body',
              'Garment and Shoe care',
              'Items',
              'Nightwear',
              'Shoes',
              'Socks & Tights',
              'Stationery',
              'Swimwear',
              'Underwear',
              'Underwear/nightwear',
              'Unknown',
              'A',
              'B',
              'C',
              'D',
              'F',
              'G',
              'H',
              'I',
              'J',
              'S']

### Scale the Training Dataset 

In [86]:
def scale_trainset_features_2(df, dtype, no_scale):
    for x in df.columns[df.dtypes == dtype]:
        if x != no_scale:
            df[x] = scaler.fit_transform(np.array(df[x]).reshape(-1,1))    

In [87]:
no_scale = 'article_id'
scale_dtypes = ['int64','int32','float64']
#for data_types in scale_dtypes:
#    scale_trainset_features(train, data_types, no_scale)

In [88]:
train.head()

t_dat                                        customer_id  \
134705 2020-07-16  bc278ea87e8891c9337a012fce80c0c49b660eb84bfbcc...   
255931 2019-06-04  1751df5330b9d5a270c936232e524ddbc8f7e3192c08ae...   
10837  2020-02-18  2574f05cc73eb0cd2d9e813c2008363e0424dfa5ca743c...   
272598 2019-04-24  d36834d69219148fe281b4c9d6ffbb977668b1bcc488e6...   
153475 2019-04-13  511ae31df843efe116ef54945214731c3165aad2f22190...   

        article_id     price  sales_channel_id  txn_year   txn_mth   txn_day  \
134705   850176003  0.059947               1.0       1.0  0.545455  0.500000   
255931   513373010  0.036530               1.0       0.5  0.454545  0.100000   
10837    759482001  0.033185               1.0       1.0  0.090909  0.566667   
272598   739819002  0.066638               1.0       0.5  0.272727  0.766667   
153475   673930003  0.033185               1.0       0.5  0.272727  0.400000   

             age  ACTIVE  ...  A  B  C  D  F  G  H  I  J  S  
134705  0.421687       1  ...  0  1  0  0  0  0  0  0  0  0  
255931  0.433735       1  ...  1  0  0  0  0  0  0  0  0  0  
10837   0.289157       1  ...  0  0  0  1  0  0  0  0  0  0  
272598  0.530120       1  ...  1  0  0  0  0  0  0  0  0  0  
153475  0.144578       1  ...  0  0  0  1  0  0  0  0  0  0  

[5 rows x 53 columns]

In [89]:
test.head()

t_dat                                        customer_id  \
72997  2018-11-26  17a4fbba8d8b728947c3d966e134ceaa5248be4bf8b88b...   
274139 2020-04-11  e1b69905e60da797f8d92946aa66e95fe6b46dedbd8bcf...   
271987 2018-12-19  ea82914cbb33f658f1cd76af29e7fe048f988302201be1...   
66429  2020-04-14  b64af996030aaa96c2e0b8228150decdf4623c90d6598e...   
132823 2018-10-24  90ae758dbc94409920a2823f14674fa430a08d8faa76bf...   

        article_id     price  sales_channel_id  txn_year   txn_mth   txn_day  \
72997    448515027  0.049911               1.0       0.0  0.909091  0.833333   
274139   803070001  0.033553               1.0       1.0  0.272727  0.333333   
271987   574750002  0.046566               1.0       0.0  1.000000  0.600000   
66429    547780031  0.049911               1.0       1.0  0.272727  0.433333   
132823   678073001  0.049911               1.0       0.0  0.818182  0.766667   

             age  ACTIVE  ...  A  B  C  D  F  G  H  I  J  S  
72997   0.072289       1  ...  0  0  0  1  0  0  0  0  0  0  
274139  0.108434       1  ...  0  1  0  0  0  0  0  0  0  0  
271987  0.469880       1  ...  0  0  1  0  0  0  0  0  0  0  
66429   0.156627       1  ...  0  0  0  1  0  0  0  0  0  0  
132823  0.409639       1  ...  0  0  0  1  0  0  0  0  0  0  

[5 rows x 53 columns]

###### 

###### 

###### 

###### 

## K Nearest Neighbour Operations

### Create Functions for NN Analysis

#### Set up function to Build Nearest Neighbours Model 

In [90]:
# Set up Unsupervised learner for implementing neighbour searches of Customer Transactions
# Set default parameters for NN algorithm - these can be overwritten during optimisationphase
def setup_NN_model(df_train, NN_metric = 'cosine', NN_algo = 'brute'):
       
    # Set up numerical data inputs for NearestNeighbors function
    df_cust_articles = df_train[nn_cols]
    
    # Set up parameters of NearestNeighbors function
    model_knn = NearestNeighbors(metric = NN_metric, algorithm = NN_algo)
    
    # Implement NearestNeighbors function on merged Trxn/Customer/Article data elements
    model_knn.fit(df_cust_articles)
    
    return model_knn, df_cust_articles

#### Set up function to  Nearest Neighbours (+ Articles bought) for a Customer 

In [91]:
# Use a function to return a list of Nearest Neighbours for a given customer.
# Return Neighbour Customer Ids and the articles they both
def get_CustSub_NeighbourIDs_articles(model_knn, df_cust_articles, df_train, NN_Neighbours):
     
    
    # Return index values for nearest neighbours - determined by 'Num_Neighbour' parameter
    distances, indices = model_knn.kneighbors(df_cust_articles.iloc[0].values.reshape(1, -1), 
                                              n_neighbors = NN_Neighbours)
     
    #print('get_CustSub_NeighbourIDs_articles\n')
    #print('get_CustSub_NeighbourIDs_articles - Indices: {}'.format(indices))
    
    # Using the Index values returned by the Nearest neighbour model
    # Generate a list of Customer Ids and Article Ids for these neighbours
    for i in range(0, len(distances.flatten())):
        if i == 0:
            customer_list = []
            article_list  = []
        else:
            customer_list.append(df_train.loc[indices.flatten()[i],'customer_id'])
            article_list.append(df_train.loc[indices.flatten()[i],'article_id'])
        
    df_neighbourhood = pd.DataFrame({'customer_id': customer_list, 'article_id': article_list})
    del customer_list, article_list
    
    return df_neighbourhood  

#### Find Top Selling Products in 'Neighbourhood' for a Customer 

In [92]:
# Function to return Top 12 Articles Purchased by the Customer 'Neighbourhood' Or Actual Data on Purchases
def get_Top12_Articles_for_Cust(df, num_of_articles):
    
    #print('Display Total Cust Rec..')
    #display(df)
      
    # Finding number of times a particular item is purchased by customers in the neighbourhood
    temp = df.groupby(['article_id'])['customer_id'].agg('count').reset_index() 
    temp.columns = ['article_id','count']
    
    # Sort the articles by the number of time purchased by the 'neighbourhood' customers
    temp = temp.sort_values('count', ascending=False)
    
    # Select the Top 12 - if there are duplicate just select the first entry
    temp = temp.nlargest(num_of_articles,'count', keep='first')
    
    # Isolate the Article Ids
    temp = temp['article_id']
    
    # Convert to dataframe and reset index
    df_article_list = pd.DataFrame({'article_id': temp})
    df_article_list = df_article_list.reset_index(drop=True)
    
    
    # Return the list of Top 12 articles
    return df_article_list

#### Invoke Function to Build KNN Model  

In [93]:
# Reset Training dataframe index - remove old one; this allows for extraction of correct index records
# returned by the NearestNeighbors function
train.reset_index(drop=True,inplace=True)

In [94]:
# Call function to set up NN learner
hm_model_knn, df_cust_articles_nn = setup_NN_model(train, NN_metric = 'cosine', NN_algo = 'brute')

###### 

###### 

###### 

## Evaluate Model

#### The code segments below implement functions to extract predictions and evaluate results.  <a class="tocSkip">

### Define Functions to Extract Customer IDs and Article Predictions from Test Data

#### Define Function to read Customer IDs from Submission File

In [95]:
# Function to return Customer Ids from sample submission file in sequence
def get_next_SubmissionCustomer(idx):
           
    # Get Customer Id from Sample Sub
    cust_id       = df_sub_sample.iloc[idx,0]
    
    df_cust_id = df_sub_sample.head(idx+1)
    df_cust_id = df_cust_id.drop(['prediction'], axis = 1)
    
    
    
    # Return index of customer id in training dataframe and the customer id itself
    return df_cust_id, cust_id

#### Define Function to Obtain Customer Transactions

In [96]:
def find_latest_CustomerTrxn(rec):
    
    # Get Next Customer Id from Submission Table
    df_one_cust, cust_id = get_next_SubmissionCustomer(rec)

    
    # Merge Customer atributes from the Customer dataset based on a join with the 'customer_id' field
    df_SubCust_Trxn = pd.merge(df_one_cust, df_trxns, how="left", on=["customer_id"])

    
    # On rare occassion that a customer has no transactions - create a dummy records to avoid a failure
    if df_SubCust_Trxn.shape[0] == 0:
        print('NO Transactions!')
        # Create dummy dataframe
        df_Dummy = pd.DataFrame([[cust_id,
                '2016-10-25',
                 651329001,
                 0.021593,
                 1]], columns=['customer_id','t_dat','article_id','price','sales_channel_id'])
        # Stack the DataFrames on top of each other
        df_SubCust_Trxn = pd.concat([df_SubCust_Trxn, df_Dummy], axis=0)
        df_SubCust_Trxn.reset_index(drop=True, inplace=True)
    
    
    # Use last transaction
    df_SubCust_Trxn1 = df_SubCust_Trxn.sort_values(by=['t_dat'], ascending=False).head(1)
    
    
    
    return df_SubCust_Trxn1, cust_id

#### Define Function to Obtain Customer Features

In [97]:
def find_latest_CustomerAttrs(df):

    
    # Merge Customer atributes from the Customer dataset based on a join with the 'customer_id' field
    df = pd.merge(df, df_customers, how="left", on=["customer_id"])
    
    # Filter out other Customer specific attributes apart from; age + fashion_news_frequency + club_member_status + postal code
    # There are very many missing values in the other Customer dataset columns, so these will be ignored
    df = df.iloc[:, [0, 1, 2, 3, 4, 7, 8, 9, 10]]

    
    
    return df

#### Define Function to Obtain Article Features

In [98]:
def find_latest_ArticleAttrs(df):
    
    # Merge Articles atributes from the Articles dataset based on a join with the 'article_id' field
    df = pd.merge(df, df_Articles, how="left", on=["article_id"])
     
    
    # Remove 'redundant' categorical columns from Articles (sample) dataframe
    # Consider a categorical column Redundant when is 'partnered' with a numerical code    
    df.drop(redundo_catcols_to_drop, axis=1, inplace=True)

    # Encode Categorical Values
    for column_name in article_cols_to_encode:
        df = encode_Categor_Features(df, column_name)
    
    
    return df    

#### Define Function to Obtain Article Predictions

In [99]:
def get_ArticlePredictions(knn_mdl, df, iNum_of_neighbours, TopN):
        
    #print('get_ArticlePredictions')
    #print('display(df)')
    #display(df)
    #display(df.iloc[0].values.reshape(1, -1))
    
    # Call function to find nearest neighbour Customer Ids and Article Ids
    # Pass in index, parameters for Nearest Neighbour function, and the number of neighbours to retrun
    df_neighbours = get_CustSub_NeighbourIDs_articles(knn_mdl, df, train, iNum_of_neighbours)
    
    # Call function to return list of the most popular Articles in terms of the 
    # purchases made by the customers neighbourhood
    df_cust_preds = get_Top12_Articles_for_Cust(df_neighbours, TopN)

    
    return df_cust_preds    

#### Define Functions to Parse Article Predictions for Kaggle Submission

In [100]:
# Function to return Article Id Predictions in String format required for Kaggle competition
def get_str_Article_Predictions(df):
    
    article_pred = df['article_id'].values.tolist()
    
    article_pred = ['0' + str(article_id) for article_id in article_pred]
    article_pred_str =  ' '.join(article_pred)
    
    str_article_preds = article_pred_str
    
    return str_article_preds

In [101]:
# Function to Build up dataframe for submission
def build_up_submission_df(cust, str_preds, custid_str, last_col):
    
    lst = [[cust, str_preds]]    
    #df = pd.DataFrame(lst, columns =['customer_id', 'prediction'])
    df = pd.DataFrame(lst, columns =[custid_str, last_col])
    final_df =df
    
    
    return final_df

### Define Functions to Evaluate Model Predictions aginst Test Data

In [102]:
# Set up column layout for dataframe that will score match accuracy of the KNN model against Test Data
eval_Cols = ['customer_id', 'prediction_accuracy','Num_Actual_Tes_Case_Purchases','Num_Predicted_Tes_Case_Purchases']

#### Set up function to return latest Customer transactions in the Test Data 

In [103]:
test.sort_values('t_dat').groupby('customer_id').head(1)

t_dat                                        customer_id  \
87798  2018-09-20  e1bbe09449e48336fc27a367a6c49e6fe5cc36fc6582b7...   
154549 2018-09-20  d5f16f7161f714012be248fcc52eba17cd84e7cd1812be...   
25232  2018-09-20  b589dc1ce100625b7eab6c6027e188910ded28c220a8b8...   
34728  2018-09-20  be25ffaad261d401d43b84102039e9dbf8c9dccfa072ea...   
243211 2018-09-20  3dcec9f4b8ba9fd63551467d61f735d8f056caf0b9c448...   
...           ...                                                ...   
32025  2020-09-22  0003e867a930d0d6842f923d6ba7c9b77aba33fe2a0fbf...   
157108 2020-09-22  d0a6b1bedc98d022fe8a3827c5306e798bcdad9ef0098c...   
183624 2020-09-22  d87d620c40da611850c55609e35826f33a9088ab744eea...   
55653  2020-09-22  b79a4aac4eeac52ad03c9037edc8b68542b2bc61979c76...   
293947 2020-09-22  5036c47a0b9002690d93e1ed295e1d5ae43f2c3221325b...   

        article_id     price  sales_channel_id  txn_year   txn_mth   txn_day  \
87798    237347004  0.066638               1.0       0.0  0.727273  0.633333   
154549   549804002  0.019804               1.0       0.0  0.727273  0.633333   
25232    670698003  0.075001               1.0       0.0  0.727273  0.633333   
34728    574109013  0.049911               1.0       0.0  0.727273  0.633333   
243211   690926001  0.066638               0.0       0.0  0.727273  0.633333   
...            ...       ...               ...       ...       ...       ...   
32025    827487003  0.083364               1.0       1.0  0.727273  0.700000   
157108   793506028  0.048373               1.0       1.0  0.727273  0.700000   
183624   763727003  0.042284               1.0       1.0  0.727273  0.700000   
55653    902161006  0.066638               1.0       1.0  0.727273  0.700000   
293947   351484002  0.019804               0.0       1.0  0.727273  0.700000   

             age  ACTIVE  ...  A  B  C  D  F  G  H  I  J  S  
87798   0.108434       1  ...  1  0  0  0  0  0  0  0  0  0  
154549  0.277108       1  ...  0  0  0  0  0  0  1  0  0  0  
25232   0.096386       1  ...  0  0  0  1  0  0  0  0  0  0  
34728   0.060241       1  ...  0  0  0  1  0  0  0  0  0  0  
243211  0.096386       1  ...  1  0  0  0  0  0  0  0  0  0  
...          ...     ...  ... .. .. .. .. .. .. .. .. .. ..  
32025   0.204819       1  ...  0  0  0  1  0  0  0  0  0  0  
157108  0.144578       1  ...  0  0  0  0  1  0  0  0  0  0  
183624  0.108434       1  ...  1  0  0  0  0  0  0  0  0  0  
55653   0.048193       1  ...  0  0  0  1  0  0  0  0  0  0  
293947  0.168675       1  ...  0  1  0  0  0  0  0  0  0  0  

[83722 rows x 53 columns]

In [104]:
# Get list of unique customers in Test Data
def get_Latest_CustTxn_in_TestData():
        
    df = test.sort_values('t_dat').groupby('customer_id').head(1)
    
    return df

#### Define Function to Return Evaluation Score

In [105]:
def return_Cust_Prediction_Score(cust_id, df_pred, df_actual, df_score):
    #print('Retrun Customer Prediction Score...')
    
    #df_match = df_pred.compare(df_actual, keep_equal=True)
    

    #"""Returns just the rows from the new dataframe that differ from the source dataframe"""
    merged_df = df_pred.merge(df_actual, indicator=True, how='outer')
    changed_rows_df = merged_df[merged_df['_merge'] == 'left_only']
    df_no_pred_match = changed_rows_df.drop('_merge', axis=1)
    
    #print('\ndf_no_pred_match...')
    #display(df_no_pred_match)
    
    scr_pct = (len(df_pred) - len(df_no_pred_match))/len(df_pred)
    
    lst = [[cust_id, scr_pct, len(df_actual), len(df_pred)]]    
    df = pd.DataFrame(lst, columns = eval_Cols)
    
    # df_score = df
    
    # Stack the DataFrames on top of each other
    # df_Submission_Stack = pd.concat([df_Submission_Stack, df_NextCust_Preds], axis=0)
    df_score = pd.concat([df_score, df], axis=0)
    
    
    return df_score

#### Iterate Through Test Data and Generate Predictions

In [106]:
# Creating an empty Dataframe with column names only - will be filled with each predicton evaluation
df_ModelEvalResults_Stack = pd.DataFrame(columns= eval_Cols)

In [107]:
## Work Through Test Data to obtain prediction list of Articles
customer_records = 100

df_Testdata_Customers_LastTrxn = get_Latest_CustTxn_in_TestData()
[print('Number of customers in Test Data is: {}\n'.format(len(df_Testdata_Customers_LastTrxn)))]

# Generate article predictions for each customer write to a new submission file
for cust_rec in range(0, customer_records):
    
    #print('cust_rec is...  {}\n'.format(df_Testdata_Customers_LastTrxn.iloc[cust_rec]))
       
    index_list = [cust_rec]
    df_CustRec = df_Testdata_Customers_LastTrxn.loc[df_Testdata_Customers_LastTrxn.index[index_list]]      
    df_CustTestNNVectors  = df_CustRec[nn_cols]
    
    
    C_Id = df_CustRec.iloc[0]['customer_id']
    
    df_CustRecId = df_CustRec[['customer_id']].copy()
        
    
    #print('Display: C_Id')
    #display(C_Id)
    
    #print('Display: df_CustRec')
    #display(df_CustRec)
    
    #print('Display: df_CustRecId')
    #display(df_CustRecId)
    

    #for data_types in scale_dtypes:
    #    scale_trainset_features(train, data_types, no_scale)
    
    
    #print('display(df_CustTestNNVectors)')
    #display(df_CustTestNNVectors)
    #display(df_CustTestNNVectors.iloc[0].values.reshape(1, -1))
    
    iNeighbours       = 85 # Number of Neighbours parameter for KNN model
    TopN              = 12  # Number of Artcile predictions to return
    
    df_TestCust_Preds     = get_ArticlePredictions(hm_model_knn, df_CustTestNNVectors, iNeighbours, TopN)
    
        
    # Create a dataframe of all transactions by the customer in the Test Data
    df_test = test.loc[test["customer_id"] == C_Id]
    #print('Display ...loc condition dataframe of all transactions by the customer in the Test Data')
    #display(df_test)
    
    
    df_TestCust_ActualPurchases = get_Top12_Articles_for_Cust(df_test, TopN)
    
    #print('display(df_TestCust_Preds)')
    #display(df_TestCust_Preds)
    #print('display(df_TestCust_ActualPurchases)')
    #display(df_TestCust_ActualPurchases)
    
    #for aa in range(0, len(df_TestCust_ActualPurchases)):
    for aa in range(0, 0):  # Skip for now
        p = df_TestCust_ActualPurchases.iloc[aa]['article_id']
        print('Number: {} p: ACTUAL article_id: {}\n'.format(aa,p))
        act_item    = df_Articles.loc[df_Articles['article_id'].eq(p), 'detail_desc']
        prdnm_item  = df_Articles.loc[df_Articles['article_id'].eq(p), 'product_group_name']
        grm_gnm_item  = df_Articles.loc[df_Articles['article_id'].eq(p), 'garment_group_name']
        indxnm_item = df_Articles.loc[df_Articles['article_id'].eq(p), 'index_name']
        print('Pred Item: Item: Prod Name {} {} {} {} \n'.format(prdnm_item, indxnm_item, grm_gnm_item, act_item))
    
    
    #for pa in range(0, len(df_TestCust_Preds)):
    for pa in range(0, 0): # Skip for now
        p = df_TestCust_Preds.iloc[pa]['article_id']
        print('Number: {} p: PREDICTED article_id: {}\n'.format(pa,p))
        pred_item = df_Articles.loc[df_Articles['article_id'].eq(p), 'detail_desc']
        pr_prdnm_item  = df_Articles.loc[df_Articles['article_id'].eq(p), 'product_group_name']
        pr_grm_gnm_item  = df_Articles.loc[df_Articles['article_id'].eq(p), 'garment_group_name']
        pr_indxnm_item = df_Articles.loc[df_Articles['article_id'].eq(p), 'index_name']
        print('Pred Item: Prod Name {} {} {} {}\n'.format(pr_prdnm_item, pr_indxnm_item, pr_grm_gnm_item, pred_item))
    

        
    
    df_ModelEvalResults_Stack = return_Cust_Prediction_Score(df_CustRecId['customer_id'], 
                                                             df_TestCust_Preds, 
                                                             df_TestCust_ActualPurchases,
                                                             df_ModelEvalResults_Stack)
        

Number of customers in Test Data is: 83722



In [108]:
df_ModelEvalResults_Stack.reset_index(drop=True,inplace=True)

In [109]:
print('Show Results scores..')
display(df_ModelEvalResults_Stack)

Show Results scores..


customer_id  prediction_accuracy  \
0   87798    e1bbe09449e48336fc27a367a6c49e6fe5cc3...             0.125000   
1   154549    d5f16f7161f714012be248fcc52eba17cd84...             0.000000   
2   25232    b589dc1ce100625b7eab6c6027e188910ded2...             0.083333   
3   34728    be25ffaad261d401d43b84102039e9dbf8c9d...             0.000000   
4   243211    3dcec9f4b8ba9fd63551467d61f735d8f056...             0.000000   
..                                                ...                  ...   
95  68200    bd9149a2c555afa64973ee7d823ddc3f37380...             0.083333   
96  264870    2ff5477788f963c6197c970b7d5b98b9e7e2...             0.000000   
97  143687    73e34b5c294e1d49e62679cdddb4c44ac4da...             0.000000   
98  168051    8b6a369b52bc4f6cb87c1855252ad4083e8a...             0.000000   
99  181349    3711b58f5c4fa8297e389722e8ac72f22ae9...             0.000000   

   Num_Actual_Tes_Case_Purchases Num_Predicted_Tes_Case_Purchases  
0                              1                                8  
1                              1                               12  
2                              3                               12  
3                              2                               12  
4                              1                               12  
..                           ...                              ...  
95                             1                               12  
96                             1                               12  
97                             1                               12  
98                             1                               12  
99                             1                               12  

[100 rows x 4 columns]

In [123]:
display(df_ModelEvalResults_Stack.sort_values('prediction_accuracy', ascending=False))

customer_id  prediction_accuracy  \
20  136906    0aaa9683d5f45b85d0f9a81e2f4d4ef774fe...             0.166667   
0   87798    e1bbe09449e48336fc27a367a6c49e6fe5cc3...             0.125000   
27  79744    a81dbea670c04e5a9dbe45475415d9ae13ca3...             0.100000   
43  225234    fbdf936ade6c975784502352b5eaf7a49ce4...             0.090909   
52  27935    d67d1b2ff71565d38f4d931d9fea105bc00ab...             0.090909   
..                                                ...                  ...   
63  114342    e4f9fb11b0a48dc8c619cfed65b180d49726...             0.000000   
17  5877    cfde71242b3e8e75c811ae7710f2efffb64546...             0.000000   
65  86692    3ce080153e362acc315b5a680c218d6a44fa3...             0.000000   
66  209516    891297ee402242f14544ae7080ac68787ca1...             0.000000   
99  181349    3711b58f5c4fa8297e389722e8ac72f22ae9...             0.000000   

   Num_Actual_Tes_Case_Purchases Num_Predicted_Tes_Case_Purchases  
20                             1                                6  
0                              1                                8  
27                             1                               10  
43                             1                               11  
52                             1                               11  
..                           ...                              ...  
63                             4                               12  
17                             1                               12  
65                             2                               12  
66                             1                               12  
99                             1                               12  

[100 rows x 4 columns]

###### 

###### 

## Build Submission

#### The solution file to have 1371980 prediction rows - use Sample Submission to construct  <a class="tocSkip">

### Read Sample Submission File

In [110]:
# Read in Full Sample Submission File
df_sub  = pd.read_csv('../input/sample_submission.csv', dtype={'customer_id': 'string'})

In [111]:
# Temp code to build submission framework based on a smaller fraction of the Sample Submission file
df_sub_sample = df_sub.sample(frac=0.001, replace=False)  

In [112]:
# Temp code to check dataframe matches file output
df_sub_sample.head()

customer_id  \
32514   061b71382f9eb8889abdfde0781146c9d1721d923fbf1e...   
740440  8a34d07375390262f2202a4b9a0b6f1a50f08a4d36998f...   
440111  52455c0cf142669f3fc3b61ddd7fbfb2d50bca6ea9288e...   
565460  6984f43622a29bf966571687af45b0441f630f759daa28...   
722589  86e098e70a15ba816512ed0a15e19ab8808227abb76628...   

                                               prediction  
32514   0706016001 0706016002 0372860001 0610776002 07...  
740440  0706016001 0706016002 0372860001 0610776002 07...  
440111  0706016001 0706016002 0372860001 0610776002 07...  
565460  0706016001 0706016002 0372860001 0610776002 07...  
722589  0706016001 0706016002 0372860001 0610776002 07...

#### Record record size of file for later validation.  <a class="tocSkip">

In [113]:
numCustomers = df_sub.shape[0]

### Iterate Through Submission File to Generate Article Predictions

In [114]:
# Creating an empty Dataframe with column names only - will be filled with each Customer and their set of predictions
df_Submission_Stack = pd.DataFrame(columns=['customer_id', 'prediction'])

In [115]:
# Set up loop based on Submission file size
customer_records = 0 # Skip this processing if zero


# Generate article predictions for each customer write to a new submission file
for cust_rec in range(0, customer_records):
    
    print('cust_rec is... ' + str(cust_rec) +'\n')
    
    df_SubCust_Trxn1, cust_id  = find_latest_CustomerTrxn(cust_rec)
    
    df_SubCust_Trxn1  = find_latest_CustomerAttrs(df_SubCust_Trxn1)
    
    df_SubCust_Trxn1  = find_latest_ArticleAttrs(df_SubCust_Trxn1)
    
    df_CustNNVectors  = df_SubCust_Trxn1[nn_cols]
    
    display(df_CustNNVectors)
        
    iNeighbours       = 550 # Number of Neighbours parameter for KNN model
    TopN              = 12  # Number of Artcile predictions to return
    df_cust_preds     = get_ArticlePredictions(hm_model_knn, df_CustNNVectors, iNeighbours, TopN)
    
    # Get a string output of Customer Article Predicitons - format as required for Kaggle competition
    str_articles_predictions = get_str_Article_Predictions(df_cust_preds)
    
    # Format Customer ID and list of Article predictions - add to dataframe to be used for CSV output
    df_NextCust_Preds = build_up_submission_df(cust_id, str_articles_predictions,'customer_id','prediction')
    
    
    # Stack the DataFrames on top of each other
    df_Submission_Stack = pd.concat([df_Submission_Stack, df_NextCust_Preds], axis=0)
     
    

### Write Predictions Submission File to CSV

#### SetUp Disk Directory for Submission File

In [116]:
# Set up folder for submission file with predictions
submission_path = '../submission'

In [117]:
# Create folder above project directory
os.makedirs(submission_path, exist_ok=True)  

#### Create Submission .CSV File

In [118]:
df_Submission_Stack = df_Submission_Stack.reset_index(drop=True)

In [119]:
# Write pre-test submission file to disk
df_Submission_Stack.to_csv(submission_path+'/submission_pretest.csv',index=False)

In [120]:
dfTempFinal  = pd.read_csv(submission_path+'/submission_pretest.csv')

[print("Matching the number of dfSub rows to the correct submission size. \nSubmission Row: {} v Expected Rows: {}"
                                                                                   .format(dfTempFinal.shape[0],numCustomers))]

Matching the number of dfSub rows to the correct submission size. 
Submission Row: 0 v Expected Rows: 1371980


[None]

In [121]:
# Check format of final submission file
dfTempFinal.head()

Empty DataFrame
Columns: [customer_id, prediction]
Index: []

In [122]:
# Write final submission file to disk
dfTempFinal.to_csv(submission_path+'/submission.csv',index=False)

###### 

###### 

## References

#### <a class="tocSkip">